src/block_disagg/block_disagg_addr.c

```diff
@@ -8,32 +8,118 @@
+{
+    uint8_t *p;
+    p = *pp;
+
+    WT_SIZE_CHECK_PACK(1, maxlen);
+    *p++ = WT_BLOCK_DISAGG_ADDR_VERSION | (WT_BLOCK_DISAGG_ADDR_VERSION_MIN << 4);
+
+    *pp = p;
+    return (0);
+}
+
+/*
+ * __block_disagg_addr_unpack_version --
+ *     Unpack the address cookie version from the buffer.
+ */
+static inline int
+__block_disagg_addr_unpack_version(
+  const uint8_t **pp, size_t maxlen, uint8_t *version, uint8_t *version_min)
+{
+    uint8_t version_byte;
+    const uint8_t *p;
+    p = *pp;
+
+    WT_SIZE_CHECK_UNPACK(1, maxlen);
+    version_byte = *p++;
+    *version = version_byte & 0x0f;
+    *version_min = version_byte >> 4;
+
+    *pp += 1;
+    return (0);
+}
+
+/*
+ * __block_disagg_addr_pack_uint32 --
+ *     Pack a 32-bit unsigned integer into the address cookie.
+ */
+static inline int
+__block_disagg_addr_pack_uint32(uint8_t **pp, size_t maxlen, uint32_t value)
+{
+    uint8_t *p;
+    p = *pp;
+
+    WT_SIZE_CHECK_PACK(4, maxlen);
+    memcpy(p, &value, sizeof(value));
+    p += sizeof(value);
+
+    *pp = p;
+    return (0);
+}
+
+/*
+ * __block_disagg_addr_unpack_uint32 --
+ *     Unpack a 32-bit unsigned integer from the address cookie.
+ */
+static inline int
+__block_disagg_addr_unpack_uint32(const uint8_t **pp, size_t maxlen, uint32_t *valuep)
+{
+    const uint8_t *p;
+    p = *pp;
+
+    WT_SIZE_CHECK_UNPACK(4, maxlen);
+    memcpy(valuep, p, sizeof(*valuep));
+    p += sizeof(*valuep);
+
+    *pp = p;
+    return (0);
+}
+
 /*
  * __wti_block_disagg_addr_pack --
  *     Convert the filesystem components into its address cookie.
  */
 int
-__wti_block_disagg_addr_pack(
-  uint8_t **pp, uint64_t page_id, uint64_t lsn, uint32_t size, uint32_t checksum)
+__wti_block_disagg_addr_pack(uint8_t **pp, uint64_t page_id, uint64_t flags, uint64_t lsn,
+  uint64_t base_lsn, uint32_t size, uint32_t checksum)
 {
-    uint64_t c, l, p, s;
-
-    /* FIXME-WT-14644: write extensible byte */
-    /* See the comment above: this is the reverse operation. */
     if (size == 0) {
-        p = WT_BLOCK_INVALID_PAGE_ID;
-        s = c = 0;
-        l = 0;
-    } else {
-        p = page_id;
-        l = lsn;
-        s = size;
-        c = checksum;
+        page_id = WT_BLOCK_INVALID_PAGE_ID;
+        flags = 0;
+        size = checksum = 0;
+        lsn = base_lsn = 0;
     }
-    WT_RET(__wt_vpack_uint(pp, 0, p));
-    WT_RET(__wt_vpack_uint(pp, 0, l));
-    WT_RET(__wt_vpack_uint(pp, 0, s));
-    WT_RET(__wt_vpack_uint(pp, 0, c));
+
+    /* Write the address version. */
+    WT_RET(__block_disagg_addr_pack_version(pp, 0));
+
+    /* Pack the address cookie. */
+    WT_RET(__wt_vpack_uint(pp, 0, page_id));
+    WT_RET(__wt_vpack_uint(pp, 0, flags));
+    WT_RET(__wt_vpack_uint(pp, 0, lsn));
+    WT_RET(__wt_vpack_uint(pp, 0, base_lsn));
+    WT_RET(__wt_vpack_uint(pp, 0, size));
+
+    /* Pack the checksum as a fixed-length 32-bit integer. */
+    WT_RET(__block_disagg_addr_pack_uint32(pp, 0, checksum));
```

Comment:
Oh nice, 4 bytes instead of 5.  So the version byte is "free".

Reply:
Super cool to see this!